Adjust for FullAOT removal (#844)
unoplatform/Uno.Wasm.Bootstrap: doc/runtime-execution-modes.md

---
uid: Uno.Wasm.Bootstrap.Runtime.Execution
---
# Runtime Execution Modes

> The mono for WebAssembly runtime provides three execution modes, Interpreter, AOT (Ahead of Time) and Mixed Mode Interpreter/AOT.

Reviewer: ```suggestion
The mono for WebAssembly runtime provides three execution modes, Interpreter, AOT (Ahead of Time), and Mixed Mode Interpreter/AOT.
```

The execution mode can be set as follows:

```xml
<WasmShellMonoRuntimeExecutionMode>Interpreter</WasmShellMonoRuntimeExecutionMode>
```

The possible values are:

- `Interpreter` (the default mode)
- `InterpreterAndAOT`

To setup your machine to use AOT modes on Windows, you will need to install [Python from Windows Store](https://www.microsoft.com/store/productId/9P7QFQMJRFP7), or manually through [Python's official site](https://www.python.org/downloads/).

### Interpreter mode

> This mode is the slowest, but allows for great flexibility and debugging, as well as an efficient payload size.

Reviewer: ```suggestion
This mode is the slowest but allows for great flexibility and debugging, as well as an efficient payload size.
```

This mode is a hybrid between the interpreter and the AOT modes, where the interpreter is used to execute the code, but the JIT engine is used to generate some WebAssembly code on the fly. This mode is generally faster than the interpreter, but slower than the AOT mode.

To enable this mode, use the following option:

```xml
<PropertyGroup>
    <WasmShellEnableJiterpreter>true</WasmShellEnableJiterpreter>
</PropertyGroup>
```

> Additionally, some options can be used to fine tune the Jiterpreter mode, using options found [in this file](https://github.com/dotnet/runtime/blob/6a047a9aec7a36039cffac61186b04bd3f16dbe0/src/mono/mono/utils/options-def.h#L86-L114):

Reviewer: ```suggestion
Additionally, some options can be used to fine-tune the Jiterpreter mode, using options found [in this file](https://github.com/dotnet/runtime/blob/6a047a9aec7a36039cffac61186b04bd3f16dbe0/src/mono/mono/utils/options-def.h#L86-L114):
```

Finally, runtime statistics are maintained by the jiterpreter and can be displayed by running `INTERNAL.jiterpreter_dump_stats()` in the browser debugger console.

### Mixed Interpreter and AOT Mode

> This mode enable AOT compilation for most of the assemblies, with [some specific exceptions](https://github.com/dotnet/runtime/issues/50609).

Reviewer: ```suggestion
This mode enables AOT compilation for most of the assemblies, with [some specific exceptions](https://github.com/dotnet/runtime/issues/50609).
```

> [!IMPORTANT]

> > This mode are not supported on macOS. You'll need to use a [Linux container](https://hub.docker.com/r/unoplatform/wasm-build) to build with AOT, see below for more details.

Reviewer: ```suggestion
> This mode is not supported on macOS. You'll need to use a [Linux container](https://hub.docker.com/r/unoplatform/wasm-build) to build with AOT, see below for more details.
```

The easiest is to build using the environment provided by the [unoplatform/wasm-build docker image](https://hub.docker.com/r/unoplatform/wasm-build).

## Profile Guided AOT

This mode allows for the AOT engine to selectively optimize methods to WebAssembly, and keep the rest as interpreted. This gives a very good balance when choosing between performance and payload size. It also has the advantage of reducing the build time, as less code needs to be compiled down to WebAssembly.

This feature is used in two passes:

- The first pass needs the creation of a profiled interpreter build, which records any methods invoked during the profiling session.
- The second pass rebuilds the application using the Mixed AOT/Interpreter mode augmented by the recording created during the first pass.

> This mode gives very good results, where the RayTracer sample of this repository goes from an uncomressed size of 5.5MB to 2.9MB.

Reviewer: ```suggestion
This mode gives very good results, where the RayTracer sample of this repository goes from an uncompressed size of 5.5MB to 2.9MB.
```

To create a profiled build:

- In your Wasm csproj, add the following:

  ```xml
  <WasmShellGenerateAOTProfile>true</WasmShellGenerateAOTProfile>
  ```

- In your `LinkerConfig.xml` file, add the following:

  ```xml
  <assembly fullname="WebAssembly.Bindings" />
  ```

> - Run the application once, without the debugger (e.g. Ctrl+F5)
> - Navigate throughout the application in high usage places.
> - Once done, either:

Reviewer: ```suggestion
- Run the application once, without the debugger (e.g. Ctrl+F5)
- Navigate throughout the application in high-usage places.
- Once done, either:
```

The `MixedModeExcludedAssembly` is also used to filter the profile for assemblies, see below for more information.

> Dumping the whole list of original and filtered list is possible by adding:

Reviewer: ```suggestion
Dumping the whole list of the original and filtered lists is possible by adding:
```

```xml
<PropertyGroup>
    <WasmShellGenerateAOTProfileDebugList>true</WasmShellGenerateAOTProfileDebugList>
</PropertyGroup>
```

This will generate files named `AOTProfileDump.*.txt` in the `obj` folder for inspection.

### Mixed AOT/Interpreter Mode

> This modes allows for the WebAssembly generation of parts of the referenced assemblies, and falls back to the interpreter for code that was excluded or not known at build time.

Reviewer: ```suggestion
This mode allows for the WebAssembly generation of parts of the referenced assemblies and falls back to the interpreter for code that was excluded or not known at build time.
```

  <MonoRuntimeMixedModeExcludedAssembly Include="Newtonsoft.Json" />
</ItemGroup>
```

Adding assemblies to this list will exclude them from being compiled to WebAssembly.

### Troubleshooting Mixed AOT/Interpreter Mode

> When using the Mixed AOT/Interpreter mode, it is possible that some methods may not be compiled to WebAssembly for a variety of reasons. This can cause performance issues, as the interpreter is slower than the AOT generated code.

Reviewer: ```suggestion
When using the Mixed AOT/Interpreter mode, it is possible that some methods may not be compiled to WebAssembly for a variety of reasons. This can cause performance issues, as the interpreter is slower than the AOT-generated code.
```

> ## Required configuration for AOT, Mixed Mode or external bitcode support compilation on Windows 10

Reviewer: ```suggestion
## Required configuration for AOT, Mixed Mode, or external bitcode support compilation on Windows 10
```

### Native windows tooling

This is the default mode on Windows. It requires installing [Python from Windows Store](https://www.microsoft.com/store/productId/9P7QFQMJRFP7), or manually through [Python's official site](https://www.python.org/downloads/).

> This mode is compatible with CI servers which have Python installed by default, such as [Azure Devops Hosted Agents](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops).

Reviewer: ```suggestion
This mode is compatible with CI servers that have Python installed by default, such as [Azure Devops Hosted Agents](https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops).
```

#### Powershell setup

> The bootstrapper needs to use powershell, and configuration is needed.
> 
> You'll need to run the following command in an elevated (administrator) PowerShell prompt:

Reviewer: ```suggestion
The bootstrapper needs to use PowerShell, and configuration is needed.

You'll need to run the following command in an elevated (administrator) PowerShell prompt:
```

> The integration with WSL provides a way for using AOT, Mixed mode or external bitcode support using Windows 10.

Reviewer: ```suggestion
The integration with WSL provides a way to use AOT, Mixed mode, or external bitcode support using Windows 10.
```

This feature is active only if one of those condition is true:

> - The `WasmShellMonoRuntimeExecutionMode` property is `InterpreterAndAOT

Reviewer: ```suggestion
- The `WasmShellMonoRuntimeExecutionMode` property is `InterpreterAndAOT`
```
